src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp

```diff
@@ -0,0 +1,882 @@
+  {
+    const Mat3& R = cameraPose.getTransform().rotation();
+    poseBlock = R;
+    double* poseBlockPtr = poseBlock.data();
+
+    /*Define rotation parameterization*/
+    problem.AddParameterBlock(poseBlockPtr, 9, new SO3Parameterization);
+
+    // keep the camera extrinsics constants
+    if(cameraPose.isLocked() || isConstant || !refineRotation)
+    {
+      // set the whole parameter block as constant.
+      _statistics.addState(EParameter::POSE, EParameterState::CONSTANT);
+      problem.SetParameterBlockConstant(poseBlockPtr);
+      return;
+    }
+
+    _statistics.addState(EParameter::POSE, EParameterState::REFINED);
+  };
+
+  // setup poses data
+  for(const auto& posePair : sfmData.getPoses())
+  {
+    const IndexT poseId = posePair.first;
+    const sfmData::CameraPose& pose = posePair.second;
+
+    // skip camera pose set as Ignored in the Local strategy
+    if(getPoseState(poseId) == EParameterState::IGNORED)
+    {
+      _statistics.addState(EParameter::POSE, EParameterState::IGNORED);
+      continue;
+    }
+
+    const bool isConstant = (getPoseState(poseId) == EParameterState::CONSTANT);
+
+    addPose(pose, isConstant, _posesBlocks[poseId]);
+  }
+}
+
+void BundleAdjustmentPanoramaCeres::addIntrinsicsToProblem(const sfmData::SfMData& sfmData, BundleAdjustment::ERefineOptions refineOptions, ceres::Problem& problem)
+{
+  const std::size_t minImagesForOpticalCenter = 3;
+  
+  const bool refineIntrinsicsOpticalCenter = (refineOptions & REFINE_INTRINSICS_OPTICALCENTER_ALWAYS) || (refineOptions & REFINE_INTRINSICS_OPTICALCENTER_IF_ENOUGH_DATA);
+  const bool refineIntrinsicsFocalLength = refineOptions & REFINE_INTRINSICS_FOCAL;
+  const bool refineIntrinsicsDistortion = refineOptions & REFINE_INTRINSICS_DISTORTION;
+  const bool refineIntrinsics = refineIntrinsicsDistortion || refineIntrinsicsFocalLength || refineIntrinsicsOpticalCenter;
+
+  std::map<IndexT, std::size_t> intrinsicsUsage;
+
+  // count the number of reconstructed views per intrinsic
+  for(const auto& viewPair: sfmData.getViews())
+  {
+    const sfmData::View& view = *(viewPair.second);
+
+    if(intrinsicsUsage.find(view.getIntrinsicId()) == intrinsicsUsage.end()) {
+      intrinsicsUsage[view.getIntrinsicId()] = 0;
+    }
+
+    if(sfmData.isPoseAndIntrinsicDefined(&view)) {
+      ++intrinsicsUsage.at(view.getIntrinsicId());
+    }
+  }
+
+  for(const auto& intrinsicPair: sfmData.getIntrinsics())
+  {
+    const IndexT intrinsicId = intrinsicPair.first;
+    const auto& intrinsicPtr = intrinsicPair.second;
+    const auto usageIt = intrinsicsUsage.find(intrinsicId);
+    if(usageIt == intrinsicsUsage.end()) {
+      // if the intrinsic is never referenced by any view, skip it
+      continue;
+    }
+    const std::size_t usageCount = usageIt->second;
+
+    // do not refine an intrinsic does not used by any reconstructed view
+    if(usageCount <= 0 || getIntrinsicState(intrinsicId) == EParameterState::IGNORED)
```

Comment:
`usageCount` can't be -1. Use `UndefinedIndexT` instead

Reply:
`usageCount` is a counter: so `<= 0` can be replaced by `== 0`. It's not related to `UndefinedIndexT`.